src/deploy/mod.rs

```diff
@@ -86,78 +99,123 @@ async fn calculate_rosu_pp(
+            .combo(score.max_combo as u32)
+            .misses(score.count_misses as u32)
+            .accuracy(score.accuracy as f64)
+            .calculate();
+
+        let mut pp = round(result.pp() as f32, 2);
+        if pp.is_infinite() || pp.is_nan() {
+            pp = 0.0;
+        }
 
-    let mut stars = round(result.stars() as f32, 2);
-    if stars.is_infinite() || stars.is_nan() {
-        stars = 0.0;
+        sqlx::query(&format!("UPDATE {} SET pp = ? WHERE id = ?", scores_table))
+            .bind(pp)
+            .bind(score.id)
+            .execute(ctx.database.get().await?.deref_mut())
+            .await?;
     }
 
-    Ok(CalculateResponse { stars, pp })
+    Ok(())
 }
 
-async fn recalculate_score(score: RippleScore, ctx: Arc<Context>) -> anyhow::Result<()> {
-    let request = CalculateRequest {
-        beatmap_id: score.beatmap_id,
-        beatmap_md5: score.beatmap_md5,
-        mode: score.play_mode,
-        mods: score.mods,
-        max_combo: score.max_combo,
-        count_300: score.count_300,
-        count_100: score.count_100,
-        count_50: score.count_50,
-        miss_count: score.count_misses,
-    };
+async fn recalculate_beatmap(
+    beatmap_md5: String,
+    scores_table: &str,
+    mods_query_str: String,
+    mode: i32,
+    rx: i32,
+    ctx: Arc<Context>,
+) -> anyhow::Result<()> {
+    let scores: Vec<LightweightScore> = sqlx::query_as(&format!(
+        "SELECT s.id, s.mods, s.max_combo, s.play_mode, b.beatmap_id, s.pp, s.accuracy, s.misses_count
+        FROM {} s
+        INNER JOIN
+            beatmaps b
+            USING(beatmap_md5)
+        WHERE
+            completed IN (2, 3)
+            AND play_mode = ?
+            AND s.beatmap_md5 = ?
+            {}
+        ORDER BY pp DESC",
+        scores_table, mods_query_str,
+    ))
+    .bind(mode)
+    .bind(beatmap_md5)
+    .fetch_all(ctx.database.get().await?.deref_mut())
+    .await?;
 
-    let response = if score.mods & RX > 0 && score.play_mode == 0 {
-        calculate_special_pp(&request, ctx.clone()).await?
-    } else {
-        calculate_rosu_pp(&request, ctx.clone()).await?
-    };
+    if scores.is_empty() {
+        return Ok(());
+    }
 
-    let rx = if score.mods & RX > 0 {
-        1
-    } else if score.mods & AP > 0 {
-        2
-    } else {
-        0
-    };
+    let base_score = scores[0].clone();
 
-    let scores_table = match rx {
-        0 => "scores",
-        1 => "scores_relax",
-        2 => "scores_ap",
-        _ => unreachable!(),
-    };
+    let grouped_scores = group_scores_by_mods(scores);
 
-    sqlx::query(&format!("UPDATE {} SET pp = ? WHERE id = ?", scores_table))
-        .bind(response.pp)
-        .bind(score.id)
-        .execute(ctx.database.get().await?.deref_mut())
-        .await?;
+    let beatmap_bytes =
+        usecases::beatmaps::fetch_beatmap_osu_file(base_score.beatmap_id, ctx.clone()).await?;
+
+    let beatmap = Beatmap::from_bytes(&beatmap_bytes)?;
```

Comment:
does this actually do all of the .osu parsing work here up front? and the state of this beatmap is not mutated when used to process a perf calculation?

Reply:
yes, it does all parsing upfront. and it doesn't mutate the beatmap beyond converting to the correct mode, which is fine as all scores here will be for the same mode